fasta rework

soypat opened the pull request:
So this is **NOT** intended as a performance boost change. This is above all an API rework for ease of use and clarity.

I have implemented a fasta.Parser type. This type is intended to be the base to parse all fasta data and replace all other implementations.

#### Benefits of `Parser.ParseNext()` when compared to `XConcurrent` functions:
* Provides the bare minimum API signature to parse a single fasta genome
* It is a simple and easy to understand API.
* No ambiguities on API use.
* Takes a `io.Reader` for user liberty to use with any API or stream.
* User has more control over how many fasta genomes they want to read
* (EDIT) Also errors! ParseNext can return a very useful error with the line number on which parsing stopped 
```
goos: linux
goarch: amd64
pkg: github.com/TimothyStiles/poly/io/fasta
cpu: Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
BenchmarkFastaLegacy-8   	      55	  19968425 ns/op	17326897 B/op	  158406 allocs/op
BenchmarkParser-8        	     118	   9643441 ns/op	16795447 B/op	   58150 allocs/op
```

soypat commented on io/fasta/fasta.go:
```diff
@@ -76,6 +80,143 @@ func Parse(r io.Reader) ([]Fasta, error) {
+	if lookingForName {
+		return Fasta{}, totalRead, fmt.Errorf("did not find fasta start '>', got to line %d", p.line)
+	}
+	if lookingForName && len(sequence) == 0 {
```
Oh snap, bug! This line should read 

```go
if !lookingForName && len(sequence) == 0 {
    // We found a fasta name but no sequence to go with it.
```

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -76,6 +80,150 @@ func Parse(r io.Reader) ([]Fasta, error) {
+// NewParser returns a Parser that uses r as the source
+// from which to parse fasta formatted sequences.
+func NewParser(r io.Reader) *Parser {
+	// What's the largest line size a fasta can be?
```
I believe there's no upper limit for fasta string length but some do limit to a certain number of characters per line. 70-80 appears to be common.

https://en.wikipedia.org/wiki/FASTA_format#Original_format_&_overview

soypat replied:
Buffer could be **considerably** smaller then, heh.

TimothyStiles commented on tutorials/000_introduction_test.go:
```diff
@@ -22,7 +22,7 @@ to ship a set of real tutorials to get people started with this special
 deployable, and accessible tutorial series until today.
 
-This book is built into Poly's source code itself. Each tutorial a series
+This book is built into Poly's source code itself. Each tutorial is a series
```
I believe this was also in the last push you submitted @soypat? Do you know what's going on here?

soypat replied:
Git mystifies me. This tends to happen to me sometimes. I can try squashing commits, or that step can be done via github. As you'd prefer

TimothyStiles replied:
Whatever is easiest! @soypat

TimothyStiles commented on io/fasta/fasta_test.go:
```diff
@@ -11,6 +11,31 @@ import (
+	_ = fastas
+}
+
+func BenchmarkParser(b *testing.B) {
```
this benchmark errors out and doesn't hit any code.
![Screen Shot 2022-12-01 at 8 34 13 PM](https://user-images.githubusercontent.com/7042260/205215517-e2480367-8ce9-47d5-81bf-e5925dd3ba3f.png)

soypat replied:
Yeah, these benchmarks depended on unzipping the gz fasta, my bad. Fixing now!

TimothyStiles commented:
@soypat the benchmark needs some fixing but this is looking really good! Is your plan to eventually deprecate the legacy fasta parsers as well?

soypat commented:
@TimothyStiles I'd really like to see those "Concurrent" parsers go. Maybe `Read` can be renamed as `ReadFile`, but that'd be gravy. Parse is OK as shorthand for creating a new parser and reading all.

soypat commented:
@TimothyStiles  So I have changed the following things:
* I initialize uniprot 1MB fasta as a raw format string in TestMain. This will avoid some noisiness of file manipulation in benchmarks
* I have added an argument to NewParser: maxLineSize, in bytes. This is to give users some freedom of how much heap they want to use for the parser. 
* Added error handling during benchmarks
* Better error handling for the non-EOF case during parsing

TimothyStiles commented:
@soypat recent additions are looking great. Having a shorthand like `Parse` would be excellent.

`Reset`, `ParseAll` and `ParseN` need to be tested and it looks there are a couple of corner cases in `ParseNext` that also need to be tested.
![Screen Shot 2022-12-02 at 1 16 24 PM](https://user-images.githubusercontent.com/7042260/205387618-b339915c-c657-457a-99bf-28c237b3ef69.png)
![Screen Shot 2022-12-02 at 1 16 32 PM](https://user-images.githubusercontent.com/7042260/205387627-27f9e726-02b5-4e7a-ae9a-76e455220880.png)

soypat commented:
@TimothyStiles Alright, looks much better. Fixed various bugs and test coverage is up to 95%

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -76,6 +80,176 @@ func Parse(r io.Reader) ([]Fasta, error) {
+		}
+		fastas = append(fastas, fasta)
+	}
+	return fastas, nil
```
success case doesn't seem to be covered by tests?

TimothyStiles commented:
@soypat I just rewrote `fasta.Parse` as a wrapper around `fasta.ParseAll` and had a couple of tests fail.

`ExampleWrite` fails and doesn't return anything when trying to read what was written out. Couple of things are happening here.

1. New parser is stripping the `*` character which denotes a stop codon. Since it's technically not a amino acid it doesn't have a letter but is important to denote.
2. It looks like there's a fatal difference between what the parser can read and what the writer outputs.

![Screen Shot 2022-12-06 at 1 44 25 PM](https://user-images.githubusercontent.com/7042260/206030701-b98424f9-53fe-45d6-b26c-30ab5b2b93c8.png)
![Screen Shot 2022-12-06 at 1 44 11 PM](https://user-images.githubusercontent.com/7042260/206030715-93858130-6e8f-4921-afea-361fca80887e.png)
![Screen Shot 2022-12-06 at 1 43 57 PM](https://user-images.githubusercontent.com/7042260/206030718-eded7a08-ec08-4b53-af9a-aadfa539f06b.png)

Been super busy the past several days but will be free to work on this together any time this week after tonight!

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -66,14 +70,178 @@ type Fasta struct {
+			if errors.Is(err, io.EOF) {
+				if len(line) > 1 && line[0] != ';' {
+					// EOF-ended fasta. We append line if not empty
+					stopTranslate := bytes.IndexByte(line, translationStopChar)
```
The stop codon only denotes that a gene will stop being transcribed. In practice a single sequence can have many stop codons. Is it safe to remove this block @soypat?

soypat replied:
Oh snap, right will do. Just pushed a incorrect solution in that case

TimothyStiles commented on io/fasta/fasta_test.go:
```diff
@@ -248,3 +304,86 @@ func TestWrite_error(t *testing.T) {
+		expected []Fasta
+	}{
+		{
+			content:  ">humen\nGATTACA\nCATGAT", // EOF-ended Fasta not valid
```
we should probably be more permissive and along EOF-ended Fasta as valid?

TimothyStiles replied:
@soypat my latest changes have this entire suite of tests failing on off-by-one errors where there's always one more fasta than what is expected. Any idea what's going on?

soypat replied:
I'll take a deep dive look today!

soypat replied:
I'm hesitant to say EOF ended fastas are OK. This means you can't use io.LimitReader to read fastas at all using poly API since its use will return incomplete fastas without warning. The reason this is important to me is because io.LimitReader is a commonly used foundation for writing software resilient to adversarial user input.

Though I'd also like users who want to parse EOF ended fastas to be able to use poly. This is possible since we can return a valid fasta alongside an io.EOF error when the fasta is immediately terminated by an EOF. This will allow users who need this functionality to build their own parsers who treat the EOF case specially.

Note: ParseNext now (correctly?) parses `data/empty.fasta`, returning a single fasta.

soypat replied:
Maybe we can add a boolean flag to the ParseN, ParseAll, ParseBytes: `acceptEOFTerminatedFastas`?

TimothyStiles replied:
> Maybe we can add a boolean flag to the ParseN, ParseAll, ParseBytes: acceptEOFTerminatedFastas?

This feels a little clunky but may be the way to do it. Should this variadic or explicit?

TimothyStiles replied:
> Note: ParseNext now (correctly?) parses data/empty.fasta, returning a single fasta.

Just made empty.fasta truly empty. It's meant to test the corner cases I just commented on.

soypat replied:
> This feels a little clunky but may be the way to do it. Should this variadic or explicit?

Maybe we can have parser options? Something like this https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis? Or maybe that's too much- just a struct with fields?

soypat commented on io/fasta/fasta.go:
```diff
@@ -66,14 +70,186 @@ type Fasta struct {
+			// So got to this point the line is probably OK, we will return a Fasta.
+			// We may or may not exit with error depending on whether we find a terminator.
+			sequence = append(sequence, line...)
+			fastaIsTerminated := bytes.IndexByte(line, '*') >= 0
```
So just checking, although base.fasta uses a stop codon at the end of the fasta sequence, it could easily have more genes that follow?

TimothyStiles replied:
Yes. The Stop codon is denoting the stopped biological transcription/translation of a sequence not the computational representation of the sequence.

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -66,14 +70,181 @@ type Fasta struct {
+				break
+			} else if errors.Is(err, bufio.ErrBufferFull) {
+				// Buffer size too small to read fasta line.
+				return Fasta{}, totalRead, fmt.Errorf("line %d too large for buffer, use larger maxLineSize: %w", p.line+1, err)
```
How would we make this fail? I've tried adjusting the parser struct's max line size variable to 1 but still am not able to get this to be thrown.

soypat replied:
bufio.Reader has a minimum permissible line length of 16. Test now does that, though I feel most of this "error" coverage is in vain. This error in particular can be avoided using bufio.Reader.ReadLine instead of ReadSlice. It just makes parsing harder.

soypat replied:
That's not to say the maxLineLength argument is in vain, it still provides a way to let users avoid heap allocations during scan by setting an appropiate value.

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -66,14 +70,181 @@ type Fasta struct {
+
+	// Parsing ended. Check for inconsistencies.
+	if lookingForName {
+		return Fasta{}, totalRead, fmt.Errorf("did not find fasta start '>', got to line %d: %w", p.line, err)
```
I can't seem to reach this code while parsing with an empty fasta file. Is there a way to do this?

soypat replied:
Done!

TimothyStiles commented on io/fasta/fasta.go:
```diff
@@ -66,14 +70,181 @@ type Fasta struct {
+	if lookingForName {
+		return Fasta{}, totalRead, fmt.Errorf("did not find fasta start '>', got to line %d: %w", p.line, err)
+	}
+	if !lookingForName && len(sequence) == 0 {
```
Can't reach with test using an empty fasta file.

soypat replied:
Done!

soypat commented:
I removed `empty.fasta` since it can be represented as a const string
